enclave-runtime/src/rpc/author/top_filter.rs

```diff
@@ -0,0 +1,122 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use ita_stf::TrustedOperation;
+
+/// Trait for filtering values
+///
+/// Returns `Some` if a value should be included and `None` if discarded
+pub trait Filter {
+	type Value;
+
+	fn filter<'a>(&self, value: &'a Self::Value) -> Option<&'a Self::Value>;
+}
+
+/// Filter that allows all TOPs (i.e. not filter at all)
+pub struct AllowAllTopsFilter;
+
+impl Filter for AllowAllTopsFilter {
+	type Value = TrustedOperation;
+
+	fn filter<'a>(&self, value: &'a Self::Value) -> Option<&'a Self::Value> {
```

Comment:
The idea behind this signature is a functional chain with which you process a collection of things. So the basic idea is to do:
```
tops.iter().map(pre_process).map(filter).map(execute). ...
```
To be very specific, the filter will probably have to be called in a `flat_map` to discard all items that are `None`.

We are not there yet in our specific use case. But that's generally how filters are designed nowadays and if the calling code ever evolves to be more streamlined, this signature is what we want.

Reply:
Hmm, I am not sure... borrowed values in options tend to be cumbersome to work with.

For your described use case, you could also use:

```
tops.iter().filter(filter).map(execute)
```
where the filter is a boolean predicate. I am not convinced yet of that signature. :)

Reply:
Okay also here, I probably have to eat my words again. In Rust the `iterator` has a `filter` method, which takes a `Predicate`, which is a function that returns a bool, rather than the value itself. So my whole argument is not valid anymore. I'll change it to bool then 👍 😅

Reply:
Seems like we answered at the same time @clangenb 😉 I've changed to signature now, to be compatible with `Iterator`'s filter method 👍

Reply:
Haha. I thought you were not satisfied with my explanation. 😄